draft-ietf-quic-tls.md

```diff
@@ -667,73 +600,85 @@ Early data within the TLS connection MUST NOT be used.  As it is for other TLS
+4.1.4) can be used to correct a client's incorrect KeyShare extension
+as well as for a stateless round trip check. From the perspective of
+QUIC, this just looks like additional messages carried in the Initial
+encryption level. Although it is in principle possible to use this
+feature for address verification in QUIC, QUIC implementations SHOULD
+instead use the RETRY feature ([TODO: REF]).
 
-Most key derivations in this document use the QHKDF-Expand function, which uses
-the HKDF expand function and is modelled on the HKDF-Expand-Label function from
-TLS 1.3 (see Section 7.1 of {{!TLS13}}). QHKDF-Expand differs from
-HKDF-Expand-Label in that it uses a different base label and omits the Context
-argument.
 
-~~~
-QHKDF-Expand(Secret, Label, Length) =
-   HKDF-Expand(Secret, QhkdfExpandInfo, Length)
-~~~
+## TLS Errors
 
-The HKDF-Expand function used by QHKDF-Expand uses the PRF hash function
-negotiated by TLS, except for handshake secrets and keys derived from them (see
-{{handshake-secrets}}).
+If TLS experiences an error, it MUST generate an appropriate alert
+as defined in {{TLS13}}; Section 6) and then provide it to QUIC,
+which sends the alert in a CRYPTO_CLOSE frame. All such alerts are
+"fatal".
 
-Where the `info` parameter of HKDF-Expand is an encoded `QhkdfExpandInfo`
-structure:
 
-~~~
-struct {
-   uint16 length = Length;
-   opaque label<6..255> = "QUIC " + Label;
-} QhkdfExpandInfo;
-~~~
+# QUIC Packet Protection {#packet-protection}
+
+As with TLS over TCP QUIC encrypts packets with keys derived from the TLS
+handshake, using the AEAD algorithm negotiated by TLS.
 
-For example, assuming a hash function with a 32 octet output, derivation for a
-client packet protection key would use HKDF-Expand with an `info` parameter of
-0x00200851554943206b6579.
 
+## QUIC Packet Encryption Keys {#encryption-keys}
 
-### Handshake Secrets {#handshake-secrets}
+QUIC derives packet encryption keys in the same way as TLS 1.3:
+Each encryption level/direction pair has a secret value, which
+is then used to derive the traffic keys using as described
+in {{TLS13}}; Section 7.3.
 
-Packets that carry the TLS handshake (Initial, Retry, and Handshake) are
-protected with a secret derived from the Destination Connection ID field from
-the client's Initial packet.  Specifically:
+The keys for the Initial encryption level are computed based on
+the client's first Destination Connection Id, as described in
+{{initial-secrets}}.
+
+The keys for the remaining encryption level are computed in the same
+fashion as the corresponding TLS keys (see {{TLS13}}; Section 7),
+except that the label for HKDF-Expand-Label uses the prefix "quic "
+rather than "tls 13". The purpose of this change is to provide key
+separation between TLS and QUIC, so that TLS stacks can avoid
+exposing TLS record protection keys.
+
+### Initial Secrets {#initial-secrets}
+
+Packets that carry the TLS handshake (Initial and Handshake) are protected with
```

Comment:
Wait, Handshake packets are protected with the Initial key, and not with the Handshake key?  Then what packet type uses the Handshake key?  If Handshake packets can be protected with either key, how does the receiver differentiate again?

I'm hoping the simplest answer is correct, and Handshake shouldn't be in this section.

Reply:
CH and SH are protected with the Initial key and the other messages are protected with the Handshake key. But yes, the rest of this text is wrong.